### PR TITLE
Cleaner exit when user credentials are incorrect

### LINKED ISSUE #2153: Offer opportunity to log in again when credentials are incorrect
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a request for a new feature in the Data Safe Haven or an upgrade to an existing feature.
- [x] The feature is still missing in the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :strawberry: Suggested change

If an administrator logs in using the wrong MS Graph credential during deployment, it isn't at all clear how they can correct their mistake. For the Azure admin account, it's easy - just run `az login`. But it isn't at all clear what to do for MS Graph. 

## :steam_locomotive: How could this be done?

Offer an option to log in again when the credential is incorrect, instead of just offering yes/no.

Possible simpler possibility is to tell people to delete the cached credential (or delete it for them) - see https://github.com/craddm/data-safe-haven/tree/delete-graph-cred (can PR once 5.0.0 merged with develop)

### LINKED ISSUE #2280: Misleading error messages when user uses wrong credentials
## :white_check_mark: Checklist

- [x] I have searched open and closed issues for duplicates.
- [x] This is a problem observed when **deploying** a Data Safe Haven.
- [x] I can reproduce this with the [latest version](https://github.com/alan-turing-institute/data-safe-haven/releases).
- [x] I have read through the [documentation](https://alan-turing-institute.github.io/data-safe-haven/).
- [x] This isn't an open-ended question (open a [discussion](https://github.com/alan-turing-institute/data-safe-haven/discussions) if it is).

## :computer: System information

- **Operating System:** debian
- **Data Safe Haven version:** develop

## :package: Packages

<details>
<summary>List of packages</summary>

```none
Paste list of packages here
```

</details>

## :no_entry_sign: Describe the problem

When the user is asked if their details are correct when deploying an SRE, if they say "no", the following messages state that the SHM needs to be redeployed. 

## :deciduous_tree: Log messages

image

### PR DESCRIPTION
**Pull request by craddm**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :arrow_heading_up: Summary

When the user rejects the current credentials when asked to confirm they are correct, code would still continue to run, resulting in numerous additional, unnecessary and misleading error messages.

In addition, when the user rejects the Graph API credentials as incorrect, there is currently no clear way to *change* those credentials.
Graph API log-in is, unlike Az CLI log-in, not done by the user at the command line.

This PR simply exits when the user confirms that the credentials are incorrect, and prints only informative error messages. 
In addition, it advises the user how to remove the cached Graph API credentials so as to be able to login with other credentials. 

### :closed_umbrella: Related issues

Closes #2280 
Closes #2153 

### :microscope: Tests

Tested locally

### CODE REVIEW

File: data_safe_haven/external/api/credentials.py [hunk 1]
```diff
@@ -144,8 +145,7 @@ def get_credential(self) -> TokenCredential:
             self.logger.error(
                 "Please authenticate with Azure: run '[green]az login[/]' using [bold]infrastructure administrator[/] credentials."
             )
-            msg = "Error getting account information from Azure CLI."
-            raise DataSafeHavenAzureError(msg) from exc
+            raise typer.Exit(code=1) from exc
```
Comment (JimMadge) [inline data_safe_haven/external/api/credentials.py]: I don't think we should raise this exception anywhere except from Typer `command`s.

This module shouldn't depend on Typer.

**Review by JimMadge — DISMISSED** [review]
I don't think this is how we should solve this. We want to catch the problem and raise a suitable exception, which is handled by the command function.

**Comment by JimMadge** [comment]
Example output

Screenshot 2024-11-20 at 10 02 17

**Review by JimMadge — APPROVED** [review]
@craddm Could you take a look at this and say what you think?

I can't request a review from you as you opened the PR 😆

**Comment by craddm** [comment]
LGTM

**Pull request merged**